File: 백준/Gold/1717. 집합의 표현/집합의 표현.py
import sys
read = sys.stdin.readline
sys.setrecursionlimit(100000)
n, m = map(int, read().split())
l = [0] * (n + 1)
tmp = -1

def union(x, y):
    x2 = find(x)
    y2 = find(y)
    if x2 != y2:
        l[y2] = x2

def find(z):
    if z == l[z]:
        return z
    else:
        l[z] = find(l[z])
        return l[z]

for i in range(n + 1):
    l[i] = i

for i in range(m):
    q, a, b = map(int, read().split())
    if q == 0:
        union(a, b)
    else:
        a2 = find(a)
        b2 = find(b)
        if a2 == b2:
            print('YES')
        else:
            print('NO')

File: 백준/Gold/1717. 집합의 표현/집합의 표현.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class Main {
    public static int[] parent;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        StringBuilder sb = new StringBuilder();

        int N = Integer.parseInt(st.nextToken());
        int M = Integer.parseInt(st.nextToken());

        parent = new int[N + 1];

        for (int i = 0; i <= N; i++) {
            parent[i] = i;
        }

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());

            int cmd = Integer.parseInt(st.nextToken());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            if (cmd == 0) {
                union(a, b);
            } else {
                if (find(a) == find(b)) {
                    sb.append("YES").append("\n");
                } else {
                    sb.append("NO").append("\n");
                }
            }
        }

        System.out.println(sb);
    }

    public static int find (int a) {
        if (a == parent[a]) {
            return a;
        }

        return parent[a] = find(parent[a]);
    }

    public static void union (int a, int b) {
        int parentA = find(a);
        int parentB = find(b);

        if (parentA != parentB) {
            parent[parentB] = parentA;
        }
    }
}